File: ghost_healer/framework/java/HealQA.java
package framework.java;

import org.openqa.selenium.*;
import org.openqa.selenium.support.events.EventFiringDecorator;
import org.openqa.selenium.support.events.WebDriverListener;
import java.net.HttpURLConnection;
import java.net.URL;
import java.io.OutputStream;
import java.io.InputStream;
import java.util.Scanner;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

/**
 * 👻 HealQA Ghost Engine (Java)
 * Intercepts Selenium actions and heals them transparently.
 */
public class HealQA {

    /**
     * Wraps a native driver to make it self-healing.
     */
    public static WebDriver getGhostDriver(WebDriver driver) {
        WebDriverListener listener = new HealingListener(driver);
        return new EventFiringDecorator<>(listener).decorate(driver);
    }

    private static class HealingListener implements WebDriverListener {
        private final WebDriver driver;

        public HealingListener(WebDriver driver) {
            this.driver = driver;
        }

        @Override
        public void onError(Object target, Method method, Object[] args, InvocationTargetException e) {
            // If findElement or an action fails, we trigger healing!
            System.out.println("👻 Ghost Mode: Detected failure. Healing...");

            // Logic to heal based on the failed selector...
            // (In a full implementation, we'd extract the selector from the exception)
        }
    }

    /**
     * Legacy support for those who still want an explicit call.
     */
    public static void safePerform(WebDriver driver, String selector, String action, String value) {
        try {
            WebElement el = driver.findElement(By.cssSelector(selector));
            if (action.equals("fill"))
                el.sendKeys(value);
            else
                el.click();
        } catch (Exception e) {
            String healed = callAI(driver, selector, action);
            if (healed != null) {
                WebElement el = driver.findElement(By.cssSelector(healed));
                if (action.equals("fill"))
                    el.sendKeys(value);
                else
                    el.click();
            }
        }
    }

    private static String callAI(WebDriver driver, String selector, String action) {
        try {
            String dom = (String) ((JavascriptExecutor) driver)
                    .executeScript("return document.documentElement.outerHTML");
            URL url = new URL("http://localhost:8000/api/heal-locator");
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod("POST");
            conn.setRequestProperty("Content-Type", "application/json");
            conn.setDoOutput(true);

            String json = String.format("{\"original_locator\":\"%s\", \"action\":\"%s\", \"dom_snapshot\":\"%s\"}",
                    selector, action, dom.replace("\"", "\\\"").replace("\n", " "));

            try (OutputStream os = conn.getOutputStream()) {
                os.write(json.getBytes());
            }

            if (conn.getResponseCode() == 200) {
                try (InputStream is = conn.getInputStream(); Scanner s = new Scanner(is)) {
                    String resp = s.useDelimiter("\\A").next();
                    return resp.split("\"healed_locator\": \"")[1].split("\"")[0];
                }
            }
        } catch (Exception e) {
        }
        return null;
    }
}

File: ghost_healer/framework/ts/HealQA.ts
import { By, WebDriver, WebElement, until } from 'selenium-webdriver';
import axios from 'axios';
import * as fs from 'fs';
import * as path from 'path';

/**
 * 👻 HealQA Ghost Engine (TypeScript)
 * This utility "decorates" a normal Selenium driver with AI healing.
 */
export class HealQA {
    /**
     * Wraps a native driver to make it self-healing.
     */
    static getGhostDriver(driver: WebDriver): WebDriver {
        return new Proxy(driver, {
            get(target: any, prop: string) {
                const value = target[prop];
                if (prop === 'findElement') {
                    return async (...args: any[]) => {
                        const selector = args[0].value; // Get CSS selector
                        try {
                            const el = await value.apply(target, args);
                            return HealQA.wrapElement(driver, el, selector);
                        } catch (err) {
                            return HealQA.healAndGetElement(driver, selector);
                        }
                    };
                }
                return typeof value === 'function' ? value.bind(target) : value;
            }
        });
    }

    private static wrapElement(driver: WebDriver, element: WebElement, selector: string): WebElement {
        return new Proxy(element, {
            get(target: any, prop: string) {
                const value = target[prop];
                if (['click', 'sendKeys', 'clear'].includes(prop)) {
                    return async (...args: any[]) => {
                        try {
                            return await value.apply(target, args);
                        } catch (err) {
                            console.warn(`👻 Ghost Mode: Element failure. Healing...`);
                            const healed = await HealQA.healAndGetElement(driver, selector);
                            return await (healed as any)[prop](...args);
                        }
                    };
                }
                return typeof value === 'function' ? value.bind(target) : value;
            }
        });
    }

    private static async healAndGetElement(driver: WebDriver, selector: string): Promise<WebElement> {
        // AI HEALING LOGIC
        const agentPath = path.join(__dirname, '..', '..', 'assets', 'heal_qa_agent.js');
        const agentJs = fs.readFileSync(agentPath, 'utf8');
        await driver.executeScript(agentJs);
        const snapshot: any = await driver.executeScript('return HealQA.getDomSnapshot()');

        const response = await axios.post('http://localhost:8000/api/heal-locator', {
            selector: selector,
            dom_snapshot: snapshot.html,
            page_url: snapshot.url,
            action: 'click',
            test_name: 'Ghost_Mode_TS'
        });

        const healed = response.data.healed_locator;
        if (healed) {
            await driver.executeScript(`HealQA.highlight('${healed}')`);
            return await driver.findElement(By.css(healed));
        }
        throw new Error(`[HealQA] Failed to find or heal element: ${selector}`);
    }
}
